lkmpg.tex

```diff
@@ -1502,6 +1502,43 @@ \subsection{Flashing keyboard LEDs}
 Flashing keyboard LEDs can be such a solution: It is an immediate way to attract attention or to display a status condition.
 Keyboard LEDs are present on every hardware, they are always visible, they do not need any setup, and their use is rather simple and non-intrusive, compared to writing to a tty or a file.
 
+Since v4.14, the timer API started to change. The series of improvements ended up at v4.15.
+The older version had redundant \cpp|data| field in the \cpp|timer_list| structure.
+A buffer overflow in the area of a \cpp|timer_list| structure may be able to overwrite the \cpp|function| and \cpp|data| fields, providing attackers with a way to use return-object programming to call arbitrary functions within the kernel.
+Also, the prototype of the callback, with a single \cpp|unsigned long| argument, will let any type checking won't work and it is an obstacle to the control flow integrity.
+So, the timer callback should pass a pointer to the \cpp|timer_list| structure rather than an \cpp|unsigned long| argument.
+Then, wrap all the information the callback needs, including the \cpp|timer_list| structure, for simply using the \cpp|container_of| macro instead of the \cpp|unsigned long| value.
+
+Before v4.14, we used \cpp|setup_timer| to initialize the timer and the \cpp|timer_list| structure looked like:
+\begin{code}
+struct timer_list {
+    unsigned long expires;
+    void (*function)(unsigned long);
+    unsigned long data;
+    u32 flags;
+};
+void setup_timer(struct timer_list *timer,
```

Comment:
Did you indent?

Reply:
The `timer_list` structure is used by clang-format, but I manually indent the functions. 
The use clang-format to indent the functions will be like:
```cpp
void timer_setup(struct timer_list *timer,
                 void (*callback)(struct timer_list *), unsigned int flags);
```
Should I use it?

Reply:
Yes, go ahead.